File: coding/raspberry/communication/protocol/MessageManager.cpp
#include <iostream>
#include <stdlib.h>
#include "MessageManager.h"

#include <pybind11/pybind11.h>

namespace py = pybind11;

uint8_t MessageManager::packPayload() {
  uint8_t *ptr = payload;
  
  memcpy(ptr, &mask, sizeof(mask)); ptr += sizeof(mask);
  
  if (mask & MOTOR_ENCODER_MASK) { 
    memcpy(ptr, data.motor_encoders, sizeof(uint16_t) * 4); ptr += sizeof(uint16_t) * 4; len++; 
    mask &= ~MOTOR_ENCODER_MASK;            // Clear MOTOR_ENCODER_BIT after packing
  }
  if (mask & HOME_SWITCH_MASK) { 
    memcpy(ptr, data.home_switches, sizeof(bool) * 4); ptr += sizeof(bool) * 4; len++; 
    mask &= ~HOME_SWITCH_MASK;              // Clear HOME_SWITCH_BIT after packing
  }
  if (mask & POTENTIOMETER_MASK) { 
    memcpy(ptr, data.potentiometers, sizeof(uint16_t) * 2); ptr += sizeof(uint16_t) * 2; len++; 
    mask &= ~POTENTIOMETER_MASK;            // Clear POTENTIOMETER_BIT after packing
  }
  if (mask & REF_DIODE_MASK) { 
    memcpy(ptr, &data.ref_diode, sizeof(uint16_t)); ptr += sizeof(uint16_t); len++; 
    mask &= ~REF_DIODE_MASK;                // Clear REF_DIODE_BIT after packing
  }
  if (mask & TEMP_SENSOR_MASK) { 
    memcpy(ptr, &data.temp_sensor, sizeof(float)); ptr += sizeof(float); len++;
    mask &= ~TEMP_SENSOR_MASK;              // Clear TEMP_SENSOR_BIT after packing
  }
  if (mask & IMU_MASK) { 
    memcpy(ptr, data.imu, sizeof(float) * 6); ptr += sizeof(float) * 6; len++;
    mask &= ~IMU_MASK;                      // Clear IMU_BIT after packing
  }

  return ptr - payload;
}
/*


void MessageManager::sendMessage() {
  uint8_t modified_mask = mask; // Local copy to track which fields are sent
  len = packPayload(modified_mask);

  uart.write(START_BYTE);
  uart.write(mask); 
  uart.write(len);
  uart.write(payload, len);

  uint8_t totalBytes = 1 + 1 + 1 + len;    // Start + mask + length + payload
  Serial.print("Sending total bytes: ");
  Serial.println(totalBytes);

  mask = modified_mask; // Update the class mask after sending
}
*/


void MessageManager::loadData() {
  uint8_t *ptr = payload;

  memcpy(&mask, ptr, sizeof(mask)); ptr += sizeof(mask);

  if (mask & MOTOR_ENCODER_MASK) {
    memcpy(data.motor_encoders, ptr, sizeof(uint16_t) * 4); ptr += sizeof(uint16_t) * 4;
  }
  if (mask & HOME_SWITCH_MASK) {
    memcpy(data.home_switches, ptr, sizeof(bool) * 4); ptr += sizeof(bool) * 4;
  }
  if (mask & POTENTIOMETER_MASK) {
    memcpy(data.potentiometers, ptr, sizeof(uint16_t) * 2); ptr += sizeof(uint16_t) * 2;
  }
  if (mask & REF_DIODE_MASK) {
    memcpy(&data.ref_diode, ptr, sizeof(uint16_t)); ptr += sizeof(uint16_t);
  }
  if (mask & TEMP_SENSOR_MASK) {
    memcpy(&data.temp_sensor, ptr, sizeof(float)); ptr += sizeof(float);
  }
  if (mask & IMU_MASK) {
    memcpy(data.imu, ptr, sizeof(float) * 6); ptr += sizeof(float) * 6;
  }
}
/*
void MessageManager::readMessage() {
  while (uart.available()) {
    uint8_t b = uart.read();
    switch (readState) {
      case WAIT_FOR_START:
        if (b == START_BYTE) readState = WAIT_FOR_MASK;
        break;

      case WAIT_FOR_MASK:
        mask = b;
        readState = WAIT_FOR_LENGTH;
        break;

      case WAIT_FOR_LENGTH:
        len = b;
        payloadIndex = 0;
        readState = WAIT_FOR_PAYLOAD;
        break;

      case WAIT_FOR_PAYLOAD:
        payload[payloadIndex++] = b;
        if (payloadIndex >= len) { // full packet received
          loadData();
          readState = WAIT_FOR_START;
        }
        break;
    }
  }
}
*/
static size_t payloadIndex = 0;

void MessageManager::fillPayload(uint8_t byte) {

  if (byte == END_BYTE){
    payloadIndex = 0;
    return;
  }
  
  payload[payloadIndex++] = byte;
}

void MessageManager::readDataFromString(const char* values[]) {

    uint8_t valuesIndex = 0;

    std::cout << "Mask: " << values[0] << std::endl;

    mask = static_cast<uint8_t>(atoi(values[valuesIndex++]));

    std::cout << "=========================MASK=========================: " << (int)mask << std::endl;

    if (mask & MOTOR_ENCODER_MASK) {
      for (int i = 0; i < 4; i++) {
        data.motor_encoders[i] = static_cast<uint16_t>(atoi(values[valuesIndex++]));
        std::cout << "Motor encoder[" << i << "]: " << data.motor_encoders[i] << std::endl;
      }
    }
    if (mask & HOME_SWITCH_MASK){
      for (size_t i = 0; i < 4; i++) {
        data.home_switches[i] = static_cast<bool>(atoi(values[valuesIndex++]));
        std::cout << "Home switch[" << i << "]: " << data.home_switches[i] << std::endl;
      }
    }
    if (mask & POTENTIOMETER_MASK) {
      for (int i = 0; i < 2; i++) {
        data.potentiometers[i] = static_cast<uint16_t>(atoi(values[valuesIndex++]));
        std::cout << "Potentiometer[" << i << "]: " << data.potentiometers[i] << std::endl;
      }
    } 
    if (mask & REF_DIODE_MASK) {
      data.ref_diode = static_cast<uint16_t>(atoi(values[valuesIndex++]));
      std::cout << "Ref diode: " << data.ref_diode << std::endl;
    }
    if (mask & TEMP_SENSOR_MASK) {
      data.temp_sensor = static_cast<float>(atof(values[valuesIndex++]));
      std::cout << "Temp sensor: " << data.temp_sensor << std::endl;
    }
    if (mask & IMU_MASK){
      for (size_t i = 0; i < 6; i++) {
        data.imu[i] = static_cast<float>(atof(values[valuesIndex++]));
      }
    }
}

void MessageManager::setMotorEncoder(const uint16_t motor_encoders[4]) {
  std::memcpy(data.motor_encoders, motor_encoders, sizeof(uint16_t) * 4);
  mask |= MOTOR_ENCODER_MASK;
}

void MessageManager::setHomeSwitches(const bool home_switches[4]) {
  std::memcpy(data.home_switches, home_switches, sizeof(bool) * 4);
  mask |= HOME_SWITCH_MASK;
}

void MessageManager::setPotentiometers(const uint16_t potentiometers[2]) {
  std::memcpy(data.potentiometers, potentiometers, sizeof(uint16_t) * 2);
  mask |= POTENTIOMETER_MASK;
}

void MessageManager::setRefDiode(const uint16_t ref_diode) {
  //data.ref_diode = ref_diode;
  std::memcpy(&data.ref_diode, &ref_diode, sizeof(uint16_t));
  mask |= REF_DIODE_MASK;
}

void MessageManager::setTempSensor(const float temp_sensor) {
  //data.temp_sensor = temp_sensor;
  std::memcpy(&data.temp_sensor, &temp_sensor, sizeof(float));
  mask |= TEMP_SENSOR_MASK;
}

void MessageManager::setImu(const float imu[6]) {
  std::memcpy(data.imu, imu, sizeof(float) * 6);
  /*std::cout << "IMU values in C++: ";
  for (int i = 0; i < 6; i++) { 
    std::cout << data.imu[i] << " ";
  }
  std::cout << std::endl;*/
  mask |= IMU_MASK;
}

MessageManager::Sensors& MessageManager::getSensors() {
  return data;
}

uint8_t* MessageManager::getPayload() {
    return payload;
}

File: coding/raspberry/communication/pythonprotocol/MessageManager.py
import struct

MOTOR_ENCODER_MASK = (1 << 0)
HOME_SWITCH_MASK   = (1 << 1)
POTENTIOMETER_MASK = (1 << 2)
REF_DIODE_MASK     = (1 << 3)
TEMP_SENSOR_MASK   = (1 << 4)
IMU_MASK           = (1 << 5)

WAIT_FOR_START = 1
WAIT_FOR_MASK = 2
WAIT_FOR_LENGTH = 3
WAIT_FOR_PAYLOAD = 4

START_BYTE = 0xFF

class MessageManager:

    def __init__(self):
        self.data = {
            'motor_encoders': [0] * 4,    # uint16_t[4]
            'home_switches': [0] * 4,     # uint8_t[4]
            'potentiometers': [0] * 2,    # uint16_t[2]
            'ref_diode': 0,               # uint16_t
            'temp_sensor': 0.0,           # float
            'imu': [0.0] * 6              # float[6]
        }

        self.mask = 0
        self.length = 0
        self.readState = WAIT_FOR_START
        self.payload = bytearray(64)
        self.payload_index = 0

    def PackPayloadFromData(self):
        ptr = 0

        if self.mask & MOTOR_ENCODER_MASK:
            struct.pack_into('<4H', self.payload, ptr, *self.data['motor_encoders'])
            ptr += 4 * 2

        if self.mask & HOME_SWITCH_MASK:
            struct.pack_into('<4B', self.payload, ptr, *self.data['home_switches'])
            ptr += 4 * 1

        if self.mask & POTENTIOMETER_MASK:
            struct.pack_into('<2H', self.payload, ptr, *self.data['potentiometers'])
            ptr += 2 * 2

        if self.mask & REF_DIODE_MASK:
            struct.pack_into('<H', self.payload, ptr, self.data['ref_diode'])
            ptr += 2

        if self.mask & TEMP_SENSOR_MASK:
            struct.pack_into('<f', self.payload, ptr, self.data['temp_sensor'])
            ptr += 4

        if self.mask & IMU_MASK:
            struct.pack_into('<6f', self.payload, ptr, *self.data['imu'])
            ptr += 6 * 4

        self.length = ptr

    def LoadDataFromPayload(self):
        ptr = 0
        
        if self.mask & MOTOR_ENCODER_MASK:
            self.data['motor_encoders'] = list(struct.unpack_from('<4H', self.payload, ptr))
            ptr += 4 * 2

        if self.mask & HOME_SWITCH_MASK:
            self.data['home_switches'] = list(struct.unpack_from('<4B', self.payload, ptr))
            ptr += 4 * 1

        if self.mask & POTENTIOMETER_MASK:
            self.data['potentiometers'] = list(struct.unpack_from('<2H', self.payload, ptr))
            ptr += 2 * 2

        if self.mask & REF_DIODE_MASK:
            (self.data['ref_diode']) = struct.unpack_from('<H', self.payload, ptr)
            ptr += 2

        if self.mask & TEMP_SENSOR_MASK:
            (self.data['temp_sensor']) = struct.unpack_from('<f', self.payload, ptr)
            ptr += 4

        if self.mask & IMU_MASK:
            self.data['imu'] = list(struct.unpack_from('<6f', self.payload, ptr))
            ptr += 6 * 4

    def ReadMessage(self, b): 
        b = b[0]  

        if self.readState == WAIT_FOR_START:
            if b == START_BYTE:
                self.readState = WAIT_FOR_MASK

        elif self.readState == WAIT_FOR_MASK:
            self.mask = b
            self.readState = WAIT_FOR_LENGTH

        elif self.readState == WAIT_FOR_LENGTH:
            self.length = b
            self.payload_index = 0
            self.readState = WAIT_FOR_PAYLOAD

        elif self.readState == WAIT_FOR_PAYLOAD:
            self.payload[self.payload_index] = b
            self.payload_index += 1

            if self.payload_index >= self.length:
                self.LoadDataFromPayload()
                self.readState = WAIT_FOR_START

    def ReadMessageSimulation(self, line):
        if len(line) == 0:
            return

        if self.readState == WAIT_FOR_START:
            value = int(line[0])
            if value == START_BYTE:
                self.readState = WAIT_FOR_MASK

        elif self.readState == WAIT_FOR_MASK:
            value = int(line[0])
            self.mask = value
            self.readState = WAIT_FOR_LENGTH

        elif self.readState == WAIT_FOR_LENGTH:
            value = int(line[0])
            self.length = value
            self.payload_index = 0
            self.readState = WAIT_FOR_PAYLOAD

        elif self.readState == WAIT_FOR_PAYLOAD:
            for value in line:
                value = int(value)
                self.payload[self.payload_index] = value
                self.payload_index += 1

                if self.payload_index >= self.length:   # full packet received
                    try:
                        self.LoadDataFromPayload()
                    except Exception as e:
                        print(f"Error unpacking payload in simulation: {e}")
                    self.readState = WAIT_FOR_START
    
    # SETTERS
    def SetMotorEncoders(self, motor_encoder_data) -> str: 
        self.data['motor_encoders'] = list(motor_encoder_data)
        self.mask |= MOTOR_ENCODER_MASK

    def SetHomeSwitches(self, home_switch_data):
        self.data['home_switches'] = list(home_switch_data)
        self.mask |= HOME_SWITCH_MASK

    def SetPotentiometers(self, potentiometer_data):
        self.data['potentiometers'] = list(potentiometer_data)
        self.mask |= POTENTIOMETER_MASK

    def SetRefDiode(self, ref_diode_data):
        self.data['ref_diode'] = ref_diode_data
        self.mask |= REF_DIODE_MASK

    def SetTempSensor(self, temp_sensor_data):
        self.data['temp_sensor'] = temp_sensor_data
        self.mask |= TEMP_SENSOR_MASK

    def SetIMU(self, imu_data):
        self.data['imu'] = list(imu_data)
        self.mask |= IMU_MASK